resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,165 @@
+   private var renewer: String = _
+   private var credentials: Credentials = _
+   private var tokens: Iterable[Token[_ <: TokenIdentifier]] = _
+   override def configureContainers(hadoopConfigSpec: HadoopConfigSpec): HadoopConfigSpec = {
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(submissionSparkConf)
+    logDebug(s"Hadoop Configuration: ${hadoopConf.toString}")
+    if (!UserGroupInformation.isSecurityEnabled) logError("Hadoop not configuration with Kerberos")
+    val maybeJobUserUGI =
+      for {
+        principal <- maybePrincipal
+        keytab <- maybeKeytab
+      } yield {
+        // Not necessary with [Spark-16742]
+        // Reliant on [Spark-20328] for changing to YARN principal
+        submissionSparkConf.set("spark.yarn.principal", principal)
+        submissionSparkConf.set("spark.yarn.keytab", keytab.toURI.toString)
+        logDebug("Logged into KDC with keytab using Job User UGI")
+        UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+          principal,
+          keytab.toURI.toString)
+      }
+    // In the case that keytab is not specified we will read from Local Ticket Cache
+    val jobUserUGI = maybeJobUserUGI.getOrElse(UserGroupInformation.getCurrentUser)
+    // It is necessary to run as jobUserUGI because logged in user != Current User
+    jobUserUGI.doAs(new PrivilegedExceptionAction[Void] {
+      override def run(): Void = {
+        logDebug(s"Retrieved Job User UGI: $jobUserUGI")
+        originalCredentials = jobUserUGI.getCredentials
+        logDebug(s"Original tokens: ${originalCredentials.toString}")
+        logDebug(s"All tokens: ${originalCredentials.getAllTokens}")
+        logDebug(s"All secret keys: ${originalCredentials.getAllSecretKeys}")
+        dfs = FileSystem.get(hadoopConf)
+        // This is not necessary with [Spark-20328] since we would be using
+        // Spark core providers to handle delegation token renewal
+        renewer = jobUserUGI.getShortUserName
+        logDebug(s"Renewer is: $renewer")
+        credentials = new Credentials(originalCredentials)
+        dfs.addDelegationTokens(renewer, credentials)
+        tokens = credentials.getAllTokens.asScala
+        logDebug(s"Tokens: ${credentials.toString}")
+        logDebug(s"All tokens: ${tokens.mkString(",")}")
+        logDebug(s"All secret keys: ${credentials.getAllSecretKeys}")
+        null
+      }})
+    if (tokens.isEmpty) logError("Did not obtain any Delegation Tokens")
+    val data = serialize(credentials)
+    val renewalTime = getTokenRenewalInterval(tokens, hadoopConf).getOrElse(Long.MaxValue)
+    val currentTime: Long = System.currentTimeMillis()
+    val initialTokenLabelName = s"$KERBEROS_SECRET_LABEL_PREFIX-$currentTime-$renewalTime"
+    val secretDT =
+      new SecretBuilder()
+        .withNewMetadata()
+          .withName(HADOOP_KERBEROS_SECRET_NAME)
+          .endMetadata()
+          .addToData(initialTokenLabelName, Base64.encodeBase64String(data))
```

Comment:
Hmm. The first argument for addToData is the key of a data item. This is not same as K8s resource label. Secret label exists as a child entry of the metadata. I think we should name this as data key. Otherwise, it will confuse many readers.

And we should also add the right metadata label to advertise this to the token refresh service pod. See the example in #453, quoted below. `hadoop-tokens-1503419054190-86400065` is a data item key whereas `refresh-hadoop-tokens: "yes"` is a secret key label and value. I think etcd create indexes only on metadata labels, which is important for selection filter performance:
```
$ kubectl get secret -o yaml spark.kubernetes.kerberos.dt
apiVersion: v1
data:
  hadoop-tokens-1503419054190-86400065: SERUUwABEzE5Mi4xNjguNi4xNTc6MzI2MjI1ABhraW1vb25raW1AUEVQUEVSREFUQS5DT00Ja2ltb29ua2ltAIoBXgrCSfaKAV4uzs32BAIU+m1Ar8OkaEAy0NaLlwC1/h5IaG4VSERGU19ERUxFR0FUSU9OX1RPS0VOEzE5Mi4xNjguNi4xNTc6MzI2MjIA
kind: Secret
metadata:
  creationTimestamp: 2017-08-21T23:16:34Z
  labels:
    refresh-hadoop-tokens: "yes"
  name: spark.kubernetes.kerberos.dt
  namespace: default
...
```